### PR TITLE
[InputMethodKit] Add ignored framework.

### PR DESCRIPTION
**Pull request by mandel-macaque**
The InputMethodKit framework was ignored because we did not have a
sample, but that is not longer the case. An apple sample can be found
here: https://developer.apple.com/library/archive/samplecode/NumberInput_IMKit_Sample/Introduction/Intro.html

Once the bindings are landed we can work in the sample.

The framework was never added therefore we are not changing any API from
the old code since it was never released.

Fixes: https://github.com/xamarin/xamarin-macios/issues/7487

### CODE REVIEW

File: src/inputmethodkit.cs [hunk 1]
```diff
@@ -49,43 +45,46 @@ partial interface IMKServer : IMKServerProxy {
 		[Export ("bundle")]
 		NSBundle Bundle { get; }
 
+		[Mac (10, 9)]
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]: macOS 10.9 is the earliest we support, so these attributes are not needed.

File: src/inputmethodkit.cs [hunk 2]
```diff
@@ -49,43 +45,46 @@ partial interface IMKServer : IMKServerProxy {
...
 		[Export ("inputText:client:")]
-		bool InputText (string str, NSObject sender);
+		bool GetInputText (string data, NSObject sender);
 
 		[Export ("handleEvent:client:")]
-		bool HandleEvent (NSEvent evnt, NSObject sender);
+		bool OnEvent (NSEvent mouseEvent, NSObject sender);
 
 		[Export ("didCommandBySelector:client:")]
 		bool DidCommandBySelector (Selector aSelector, NSObject sender);
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]:
```suggestion
		bool DidCommand (Selector aSelector, NSObject sender);
```

File: src/inputmethodkit.cs [hunk 3]
```diff
@@ -49,43 +45,46 @@ partial interface IMKServer : IMKServerProxy {
...
-		bool InputText (string str, nint keyCode, nuint flags, NSObject sender);
+		bool GetInputText (string data, nint keyCode, nuint flags, NSObject sender);
 
 		[Export ("inputText:client:")]
-		bool InputText (string str, NSObject sender);
+		bool GetInputText (string data, NSObject sender);
 
 		[Export ("handleEvent:client:")]
-		bool HandleEvent (NSEvent evnt, NSObject sender);
+		bool OnEvent (NSEvent mouseEvent, NSObject sender);
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]: I think `HandleEvent` works just fine in this case.

File: src/inputmethodkit.cs [hunk 4]
```diff
@@ -115,24 +114,36 @@ partial interface IMKStateSetting {
...
-	partial interface IMKMouseHandling {
+	interface IIMKMouseHandling {}
 
+	[Protocol]
+	interface IMKMouseHandling
+	{
+		[Abstract]
 		[Export ("mouseDownOnCharacterIndex:coordinate:withModifier:continueTracking:client:")]
-		bool MouseDown (nuint index, NSPoint point, nuint flags, out bool keepTracking, NSObject sender);
+		bool OnMouseDownOn (nuint index, CGPoint point, nuint flags, out bool keepTracking, NSObject sender);
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]:
```suggestion
		bool OnMouseDown (nuint index, CGPoint point, nuint flags, out bool keepTracking, NSObject sender);
```

File: src/inputmethodkit.cs [hunk 5]
```diff
@@ -115,24 +114,36 @@ partial interface IMKStateSetting {
...
 
+		[Abstract]
 		[Export ("mouseUpOnCharacterIndex:coordinate:withModifier:client:")]
-		bool MouseUp (nuint index, NSPoint point, nuint flags, NSObject sender);
+		bool OnMouseUpOn (nuint index, CGPoint point, nuint flags, NSObject sender);
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]:
```suggestion
		bool OnMouseUp (nuint index, CGPoint point, nuint flags, NSObject sender);
```

File: src/inputmethodkit.cs [hunk 6]
```diff
@@ -165,27 +176,78 @@ partial interface IMKInputController : IMKStateSetting, IMKMouseHandling {
 		NSObject Delegate { get; set; }
 
 		[Export ("server")]
-		IMKServer Server { get; }
+		IMKServer Server ();
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]: Why change it to a method?

File: src/inputmethodkit.cs [hunk 7]
```diff
@@ -165,27 +176,78 @@ partial interface IMKInputController : IMKStateSetting, IMKMouseHandling {
...
 
 		[Export ("client")]
-		NSObject Client { get; }
+		IIMKTextInput Client ();
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]: Same: why change it to a method?

File: src/inputmethodkit.cs [hunk 8]
```diff
@@ -196,13 +258,13 @@ partial interface IMKCandidates {
...
 		[Export ("show:")]
-		void Show (IMKCandidatesLocationHint locationHint);
+		void Show (IMKLocateCandidates locationHint);
 
 		[Export ("hide")]
 		void Hide ();
 
 		[Export ("isVisible")]
-		bool IsVisible { get; }
+		bool IsVisible ();
```
Comment (rolfbjarne) [inline src/inputmethodkit.cs]: Same: why change it to a method?

File: src/inputmethodkit.cs [hunk 9]
```diff
@@ -12,21 +12,15 @@
...
-		[Field ("kIMKCommandMenuItemName")]
-		NSString CommandMenuItemName { get; }
-
-		[Field ("kIMKCommandClientName")]
-		NSString CommandClientName { get; }
-	}
-
-	[Model]
-	partial interface IMKServerProxy {
+	[Protocol]
```
Comment (whitneyschmidt) [inline src/inputmethodkit.cs]: Why does this change from `Model` to `Protocol` with the change from `partial interface` to `interface`?

File: src/inputmethodkit.cs [hunk 10]
```diff
@@ -211,66 +273,97 @@ partial interface IMKCandidates {
...
+		// kVK_ANSI_S                    = 0x01,
+		// kVK_ANSI_D                    = 0x02,
+		// kVK_ANSI_F             
+		// So the array can be bound as ints 
+		[BindAs (typeof (nint []))]
 		[Export ("selectionKeys")]
-		NSObject [] SelectionKeys { get; set; }
+		NSNumber[] SelectionKeys { get;  [Export ("setSelectionKeys:")] set; } 
+
+// this are opacque pointers from CF, need to think about them
```
Comment (whitneyschmidt) [inline src/inputmethodkit.cs]: opaque

File: tests/xtro-sharpie/macOS-InputMethodKit.ignore [hunk 11]
```diff
@@ -1,83 +1,14 @@
...
-!missing-selector! IMKInputController::candidateSelectionChanged: not bound
-!missing-selector! IMKInputController::compositionAttributesAtRange: not bound
-!missing-selector! IMKInputController::doCommandBySelector:commandDictionary: not bound
-!missing-selector! IMKInputController::initWithServer:delegate:client: not bound
-!missing-selector! IMKInputController::markForStyle:atRange: not bound
-!missing-selector! IMKInputController::setDelegate: not bound
-!missing-selector! IMKServer::initWithName:bundleIdentifier: not bound
-!missing-selector! IMKServer::initWithName:controllerClass:delegateClass: not bound
+
+# this are bound as IMKServerInput informal protocol
```
Comment (whitneyschmidt) [inline tests/xtro-sharpie/macOS-InputMethodKit.ignore]: When do we use informal protocol?

File: src/inputmethodkit.cs (hunk 9 above)
Comment (rolfbjarne) [reply src/inputmethodkit.cs]: Those two changes are not related to each other.

File: tests/xtro-sharpie/macOS-InputMethodKit.ignore (hunk 11 above)
Comment (rolfbjarne) [reply tests/xtro-sharpie/macOS-InputMethodKit.ignore]: An informal protocol is the same as a category; _informal protocol_ is the old name that was used before Apple decided to call them categories (and depending on how old Apple's documentation for a particular API is, the documentation will say either _informal protocol_ (old docs) or _category_ (newer docs)).

To answer your question: it depends on how it's declared in the header.

**Comment by mandel-macaque** [comment]
Closing PR, the sample is terrible to port. The code does the bindings and I'm leaving the PR here just in case a user needs it.

**Comment by chamons** [comment]
We've decided to stop work on this PR and file it as https://github.com/xamarin/xamarin-macios/issues/7487 until we have customer interest, as the API and sample are non-trivial.

**Comment by ShikiSuen** [comment]
I was about to use CSharp to rewrite an IME project. Too bad to see this thread.

File: src/inputmethodkit.cs (hunk 7 above)
Comment (ShikiSuen) [reply src/inputmethodkit.cs:182]: On Swift you can use Client() to get the result in exactly the IMKTextInput type.

**Pull request closed without merging**